include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,338 @@
+              {
+                size_t idxP, idxQ;
+                for(int i = 0; i < population.size(); i++)
+                {
+                  if (arma::approx_equal(population[i], candidateP, "absdiff", epsilon))
+                  {
+                    idxP = i;
+                  }
+                  if (arma::approx_equal(population[i], candidateQ, "absdiff", epsilon))
+                  {
+                    idxQ = i;
+                  }
+                }
+
+                return CrowdingOperator(idxP,
+                                        idxQ,
+                                        ranks,
+                                        crowdingDistance);
+              }
+    );
+
+    // yeild new population P_{t+1}
+    Info << "NSGA2::Optimize() Get P(t+1)" << std::endl;
+    population.resize(populationSize);
+  }
+
+  std::vector<MatType> bestFront;
+
+  for(size_t f: fronts[0])
+  {
+    bestFront.push_back(population[f]);
+  }
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+  return bestFront;
+}
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType>
+inline void NSGA2::EvaluateObjectives(std::vector<MatType> population,
+                                      MultiobjectiveFunctionType objectives,
+                                      std::vector<arma::vec>& calculatedObjectives)
+{
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    calculatedObjectives[i] = objectives.Evaluate(population[i]);
+  }
+}
+
+template<typename MatType>
+inline void NSGA2::BinaryTournamentSelection(std::vector<MatType>& population)
+{
+  std::vector<MatType> children;
+
+  while (children.size() < population.size())
+  {
+    size_t indexA = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+    size_t indexB = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+
+    if (indexA == indexB)
+    {
+      if (indexB < populationSize - 1)
+        indexB++;
+      else
+        indexB--;
+    }
+
+    MatType parentA = population[indexA];
+    MatType parentB = population[indexB];
+
+    MatType childA = parentA, childB = parentB;
+
+    Info << "NSGA2::BinaryTournamentSelection() Crossover" << std::endl;
+    Crossover(childA, childB, parentA, parentB);
+
+    Info << "NSGA2::BinaryTournamentSelection() Mutate(A)" << std::endl;
+    Mutate(childA);
+    Info << "NSGA2::BinaryTournamentSelection() Mutate(B)" << std::endl;
+    Mutate(childB);
+
+    children.push_back(childA);
+    children.push_back(childB);
+  }
+
+  population.reserve(population.size() + children.size());
+  population.insert(std::end(population), std::begin(children), std::end(children));
+}
+
+template<typename MatType>
+inline void NSGA2::Crossover(MatType& childA,
+                MatType& childB,
+                MatType parentA,
+                MatType parentB)
+{
+  // crossover indices
+  auto idx = arma::randu<MatType>(childA.n_rows, childA.n_cols) < crossoverProb;
```

Comment:
`auto` is not recommended for Armadillo objects and expressions, see [here](http://arma.sourceforge.net/faq.html).

Reply:
The URL seems to be suffering from a DoS?
[This](https://web.archive.org/web/20190904040146/http://arma.sourceforge.net/faq.html) is the latest snapshot I could find.
<img width="1283" alt="Screenshot 2020-01-02 at 10 46 57 PM" src="https://user-images.githubusercontent.com/8839504/71681082-ca9d7700-2db1-11ea-848f-8bfac7d9e312.png">

Reply:
Hm, seems fine on my side, let us know if you still encounter any issues.

Reply:
It's working now! I think it was a temporary outage.